052-record-reconciled-versions.md

```diff
@@ -0,0 +1,160 @@
+
+## Current situation
+
+Currently the only way to know if an upgrade has finished to completion is to check is all the zookeeper and kafka brokers have rolled.
+https://github.com/strimzi/strimzi-kafka-operator/blob/2a1fdf9d8695bb22a2bf977b0ba5414291530207/systemtest/src/test/java/io/strimzi/systemtest/upgrade/KafkaUpgradeDowngradeST.java#L292
+Though this works, from a user standpoint it is far from ideal as it assumes a level of user knowledge about the cluster that may not be appropriate for all users.
+
+## Motivation
+
+This proposal allows a user to rely on the `Kafka` custom resources as their single source of truth, instead of watching/querying the individual kafka/zookeeper pods.
+This proposal lays out all the information we need and at what points in the reconcile. This information is already available within the Reconcile loop, so orchestrating and figuring out if a component has been fully upgraded will be propagated into the Custom Resource via status fields.
+
+## Proposal
+
+This proposal will cover the `Kafka` custom resource but could apply to other CR types that use `StrimziPodSets` such as `KafkaConnect` and `KafkaMirrorMaker2`, but could be extended to include other Custom Resources if there would be value in it.
+This proposal starts with the `Kafka` Custom Resource as the base example as it is the core component of Strimzi, and it deploys both `StrimziPodSet`s and `Deployment`s.
+
+This proposal assumes that all key value pairs mentioned will be writen to the `status` field as this is managed by the operator and can be added to without impacting end users.
+
+The following is the proposed new fields:
+
+### status.versions.reconciled
+The `status.versions.reconciled` field is patched at the end of a **successful** reconcile, and signals to a user that the operator at version `X` was able to reconcile to successful completion.
+This would be the field a user or System Test would query against to check whether upgrade was complete.
+```
+status:
+  versions:
+    reconciled: '0.37.0'
+```
+If a reconcile does not reach completion, this field is not added or updated, however in the case where it already exists it is left at the previous value.
+i.e.
+`status.versions.reconciled=0.35.0` would remain if a reconcilliation on `0.36.0` failed as the information that a prior reconcilliation passed could be useful to an end user.
+
+### status.versions.managedBy
+The `status.versions.managedBy` field is patched to the CR at the start of reconcile in the `initialStatus` and signals to a user that the operator is attempting to reconcile the CR at version `Y`, regardless of success or failure.
+```
+status:
+  versions:
+    managedBy: '0.37.0'
+```
+
```

Comment:
I have a general comment about the names for these fields. These are Strimzi operator versions, which I would make clear in naming those fields.

Reply:
Yeah the last ones are too verbose, taking `Operator` out is fine with me. But another question I have is about the higher level `versions` field which sounds to be generic. Do you envisage the usage of it for something else? Or do we need it?

Reply:
I don't think it is _needed_, I thought to better distinguish that these are version fields, and to have a base to build off of for potential future additions might be useful.
I.e. extra version metadata, though I can't give any solid examples.

I can cut out the versions field and instead simplify to:
```
status:
  strimziReconciled: 0.37.0
  strimziManagedBy: 0.37.0
```
or something else like:
status:
  reconciledVersion: 0.37.0
  managedByVersion: 0.37.0
```

Reply:
It was just a questions not a statement to remove it. I think having `versions` would allow to add more later so I would keep it. But then we need to be precise about which version we are talking about so one of your previous suggestions could be fine:

```yaml
status:
  versions:
    strimziReconciled: 0.37.0
    strimziManagedBy: 0.37.0
```

Reply:
I see this proposal being updated with using `kafka` and `lastSuccessfulReconciliationBy`. While the former sounds clear, the latter is not. Why did you decide to remove "strimzi" as part of the name?

Reply:
@ppatierno I did rewrite it to use `lastSuccessfulStrimziReconciliationBy` but I was thinking this was getting too long as a field name, so removed the `strimzi` from the name as I _felt_ it was implicit that Strimzi was reconciling it - however I can re-add it if you are in favour of it.
The objective of this field is to clear in what it is displaying to a user, and I am unsure if `lastSuccessfulStrimziReconciliationBy` is anymore clear than `lastSuccessfulReconciliationBy`, but curious on your thoughts here as we need to get the field name right!

Reply:
It is just Strimzi reconciling it. So what value does it really add to have the `strimzi` in the field name?

Reply:
> It is just Strimzi reconciling it. So what value does it really add to have the strimzi in the field name?

That's true, I was just wondering for consistency. Btw no really strong opinion I can live with the current field name.